target_api/middlewares.py

```diff
@@ -0,0 +1,42 @@
+from django.contrib.auth.models import AnonymousUser
+
+from channels.auth import AuthMiddlewareStack
+from channels.db import database_sync_to_async
+from rest_framework.authtoken.models import Token
+
+
+@database_sync_to_async
+def get_user_by_query_param(name, value):
+    try:
+        if name.lower() == 'token':
+            token = Token.objects.get(key=value)
+            user = token.user
+    except Token.DoesNotExist:
+        pass
+    return user or AnonymousUser()
+
+
+class TokenAuthMiddleware:
+    def __init__(self, inner):
+        self.inner = inner
+
+    def __call__(self, scope):
+        return TokenAuthMiddlewareInstance(scope, self)
+
+
+class TokenAuthMiddlewareInstance:
+    def __init__(self, scope, middleware):
+        self.middleware = middleware
+        self.scope = dict(scope)
+        self.inner = self.middleware.inner
+
+    async def __call__(self, receive, send):
+        params = self.scope['query_string'].decode('utf-8').split('=')
+        self.scope['user'] = await get_user_by_query_param(*params)
+
+        inner = self.inner(self.scope)
+        return await inner(receive, send)
+
+
+def TokenAuthMiddlewareStack(inner):
```

Comment:
What is the purpose of this function?

Reply:
This function is called from [here](https://github.com/afmicc/django-target-api/blob/d9355c7cfd5d41dfcc3b65aa7927bd45a1305327/target_api/routing.py#L7). It's useful to authenticate the user by the sent token